File: LeetCode/Medium/74 - Search a 2D Matrix.js
/**
 * @param {number[][]} matrix
 * @param {number} target
 * @return {boolean}
 */
var searchMatrix = function(matrix, target) {
    const n = matrix.length, m = matrix[0].length;
    
    let lo = 0, hi = n * m - 1;
    while (lo <= hi) {
        const mid = Math.floor((hi + lo) / 2),
              cur = matrix[Math.floor(mid / m)][mid % m];
        
        if (cur == target) return true;
        
        if (cur > target) hi = mid - 1;
        else lo = mid + 1;
    }
    return false;
};

File: LeetCode/Medium/74 - Search a 2D Matrix/74 - Search a 2D Matrix.cpp
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        int n = matrix.size(), m = matrix[0].size();
        int lo = 0, hi = n * m - 1, mid, cur;
        while (lo <= hi) {
            mid = lo + ((hi - lo) >> 1);
            cur = matrix[mid / m][mid % m];

            if (cur == target) 
                return true;
            if (cur > target) 
                hi = mid - 1;
            else 
                lo = mid + 1;
        }
        return false;
    }
};